crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -900,39 +921,53 @@ where
-                // Otherwise the signer must be decodable so that we can
-                // manually check the signatures
-                let public_keys_index_map =
-                    crate::account::public_keys_index_map(
-                        &self.ctx.pre(),
-                        signer,
-                    )?;
-                let threshold =
-                    crate::account::threshold(&self.ctx.pre(), signer)?
-                        .unwrap_or(1);
-                let mut gas_meter = self.ctx.gas_meter.borrow_mut();
-                tx_data
-                    .tx
-                    .verify_signatures(
-                        &[tx_data.tx.raw_header_hash()],
-                        public_keys_index_map,
-                        &Some(signer.clone()),
-                        threshold,
-                        || gas_meter.consume(crate::gas::VERIFY_TX_SIG_GAS),
-                    )
-                    .map_err(native_vp::Error::new)?;
+                // Otherwise the owner's vp must have been triggered and the
+                // relative action must have been written
+                if !verifiers.contains(signer) {
+                    let error = native_vp::Error::new_alloc(format!(
+                        "The required vp of address {signer} was not triggered"
+                    ))
+                    .into();
+                    tracing::debug!("{error}");
+                    return Err(error);
+                }
+
+                // The action is required becuse the target vp might have been
+                // triggered for other reasons but we need to signal it that it
+                // is required to validate a discrepancy in its balance change
+                // because of a masp transaction, which might require a
+                // different validation than a normal balance change
+                if !actions_authorizers.swap_remove(signer) {
+                    let error = native_vp::Error::new_alloc(format!(
+                        "The required masp authorizer action for address \
+                         {signer} is missing"
+                    ))
+                    .into();
+                    tracing::debug!("{error}");
+                    return Err(error);
+                }
             } else {
-                // We are not able to decode the signer, so just fail
+                // We are not able to decode the authorizer, so just fail
                 let error = native_vp::Error::new_const(
-                    "Unable to decode a transaction signer",
+                    "Unable to decode a transaction authorizer",
                 )
                 .into();
                 tracing::debug!("{error}");
                 return Err(error);
             }
         }
+        // The transaction shall not push masp authorizer actions that are not
+        // needed cause this might lead vps to run a wrong validation logic
+        if !actions_authorizers.is_empty() {
```

Comment:
For our VPs, that's correct, we'd just recheck the signature (which actually doesn't even happen because of the `Gadget` that caches the result of the signature check and ensures that we run that verification at most once). I'm not sure this holds for incorrectly VPs though: I believe the worst that could happen was the rejection of a valid transaction but not the acceptance of an invalid one. But some really broken VPs could actually end up accepting invalid transactions because of masp actions. So my reasoning here was to be very conservative, also because there's really no reason why someone should push a masp action which isn't required